modules/core/include/opencv2/core/quaternion.inl.hpp

```diff
@@ -866,6 +866,766 @@ Quat<T> Quat<T>::spline(const Quat<T> &q0, const Quat<T> &q1, const Quat<T> &q2,
     return squad(vec[1], s1, s2, vec[2], t, assumeUnit, QUAT_ASSUME_NOT_UNIT);
 }
 
+template <typename T>
+Quat<T> Quat<T>::createFromYRot(const T theta)
+{
+    return Quat<T>{std::cos(theta), 0, std::sin(theta), 0};
+}
+
+template <typename T>
+Quat<T> Quat<T>::createFromXRot(const T theta){
+    return Quat<T>{std::cos(theta), std::sin(theta), 0, 0};
+}
+
+template <typename T>
+Quat<T> Quat<T>::createFromZRot(const T theta){
+    return Quat<T>{std::cos(theta), 0, 0, std::sin(theta)};
+}
+
+template <typename T>
+Quat<T> Quat<T>::createFromEulerAngles(const Vec<T, 3> &angles,EulerAnglesType eulerAnglesType) {
+    Quat<T> q;
+    switch (eulerAnglesType)
+    {
+    case INT_XYZ:
+    {
+        Quat<T> q1 =  Quat<T>::createFromXRot(angles(0) / 2.);
+        Quat<T> q2 =  Quat<T>::createFromYRot(angles(1) / 2.);
+        Quat<T> q3 =  Quat<T>::createFromZRot(angles(2) / 2.);
+        q = q1 * q2 * q3;
```

Comment:
Lets reduce amount of generated binary code.

Make private function (`detail` namespace should be fine)
```
template <typename T>
Quat<T> createQuatFromRot(int axis, T angle)
{
    if (i == 0)
        return Quat<T>::createFromXRot(angle * 0.5f);
    if (i == 1)
        return Quat<T>::createFromXRot(angle * 0.5f);
    if (i == 2)
        return Quat<T>::createFromXRot(angle * 0.5f);
    CV_Assert(0);
}
```

Fill axis1,axis2,axis3 (preferable from `static const int rotationAxis[][3] = { { i1, i2, i3 }, { i1, i2, i3}, ... };` table):

```
int axis1, axis2, axis3;  // values are 0/1/2 for X/Y/Z
```

```
enum EulerAnglesType
{
    ...
#ifndef CV_DOXYGEN
    EULER_ANGLES_MAX_VALUE
#endif 
};

+ static assert for table size using EULER_ANGLES_MAX_VALUE
+ runtime assert: eulerAnglesType < EULER_ANGLES_MAX_VALUE
```

Finally, use this common code in epilog:

```
Quat<T> q1 =  Quat<T>::createQuatFromRot(axis1, angles(0));
Quat<T> q2 =  Quat<T>::createQuatFromRot(axis2, angles(1));
Quat<T> q3 =  Quat<T>::createQuatFromRot(axis3, angles(2));
q = q1 * q2 * q3;
```

Reply:
Thanks, I got it and I will update the implementation of  `toEulerAngle()` in the same way.

Reply:
@alalek The table size of rotationAxis is fixed (24,4). Is it necessary to add `static_assert`?

Reply:
Static assert is a compilation guard if someone would change/update corresponding enum values.